File: Agora-RTM-Tutorial-iOS-Objective-C/Agora-RTM-Tutorial/ImageMessageCell.m
//
//  ImageMessageCell.m
//  Agora-RTM-Tutorial
//
//  Created by SRS on 2020/4/2.
//  Copyright © 2020 Agora. All rights reserved.
//

#import "ImageMessageCell.h"
#import "AgoraRtm.h"

@interface ImageMessageCell ()
@property (weak, nonatomic) IBOutlet UILabel *leftUserLabel;
@property (weak, nonatomic) IBOutlet UIView *leftUserBgView;
@property (weak, nonatomic) IBOutlet UIImageView *leftContentImage;

@property (weak, nonatomic) IBOutlet UILabel *rightUserLabel;
@property (weak, nonatomic) IBOutlet UIView *rightUserBgView;
@property (weak, nonatomic) IBOutlet UIImageView *rightContentImage;

@property (nonatomic, assign) CellType type;
@property (nonatomic, copy) NSString *user;

@property (nonatomic, assign) long long requestId;
@property (nonatomic, copy) NSString *mediaId;
@property (nonatomic, copy) NSData *imageData;

@end

@implementation ImageMessageCell
- (void)setType:(CellType)type {
    _type = type;
    BOOL rightHidden = type == CellTypeLeft ? true : false;
    
    self.rightUserLabel.hidden = rightHidden;
    self.rightUserBgView.hidden = rightHidden;
    self.rightContentImage.hidden = rightHidden;

    self.leftUserLabel.hidden = !rightHidden;
    self.leftUserBgView.hidden = !rightHidden;
    self.leftContentImage.hidden = !rightHidden;
}

- (void)setUser:(NSString *)user {
    _user = [user copy];
    switch (self.type) {
        case CellTypeLeft: self.leftUserLabel.text = user; break;
        case CellTypeRight: self.rightUserLabel.text = user; break;
    }
}

- (void)awakeFromNib {
    [super awakeFromNib];
    self.rightUserBgView.layer.cornerRadius = 20;
    self.leftUserBgView.layer.cornerRadius = 20;
    self.requestId = 0;
}

- (void)updateType:(CellType)type message:(Message *)message{
    
    self.type = type;
    self.user = message.userId;
    
    // complete down
    if(self.mediaId != nil && [self.mediaId isEqualToString: message.mediaId] && self.imageData != nil){
        [self updateImageData:self.imageData];
        return;
    }
    
    // downing
    if(self.mediaId != nil && [self.mediaId isEqualToString: message.mediaId] && self.requestId > 0){
        return;
    }
        
    // stop download
    if(self.requestId > 0){
        [AgoraRtm.kit cancelMediaDownload:self.requestId completion:nil];
    }

    // start download
    self.requestId = 0;
    if (message.thumbnail != nil) {
        [self updateImageData:message.thumbnail];
    }
    
    if(message.mediaId != nil) {
        
        long long requestId;
        __weak ImageMessageCell *weakSelf = self;
        [AgoraRtm.kit downloadMediaToMemory:message.mediaId withRequest:&requestId completion:^(long long requestId, NSData *data, AgoraRtmDownloadMediaErrorCode errorCode) {
            
            // complete down
            weakSelf.requestId = 0;
            if(errorCode == AgoraRtmDownloadMediaErrorOk) {
                [weakSelf updateImageData:data];
            }
        }];
        
        self.requestId = requestId;
    }
    
    self.mediaId = message.mediaId;
}

- (void)updateImageData:(NSData *)imageData {

    self.imageData = imageData;

    switch (self.type) {
        case CellTypeLeft:
            self.leftContentImage.image = [UIImage imageWithData:imageData];
            self.leftContentImage.hidden = NO;
            break;
        case CellTypeRight:
            self.rightContentImage.hidden = NO;
            self.rightContentImage.image = [UIImage imageWithData:imageData];
            break;
    }
}

@end

File: Agora-RTM-Tutorial-iOS/Agora-RTM-Tutorial/ImageMessageCell.swift
//
//  ImageMessageCell.swift
//  Agora-RTM-Tutorial
//
//  Created by SRS on 2020/4/2.
//  Copyright © 2020 Agora. All rights reserved.
//

import UIKit

class ImageMessageCell: UITableViewCell {
    
    @IBOutlet weak var leftUserLabel: UILabel!
    @IBOutlet weak var leftUserBgView: UIView!
    @IBOutlet weak var leftContentImage: UIImageView!
    
    @IBOutlet weak var rightUserLabel: UILabel!
    @IBOutlet weak var rightUserBgView: UIView!
    @IBOutlet weak var rightContentImage: UIImageView!
    
    fileprivate var requestId: Int64 = 0
    fileprivate var mediaId: String?
    fileprivate var imageData: Data?
    
    override func awakeFromNib() {
        rightUserBgView.layer.cornerRadius = 20
        leftUserBgView.layer.cornerRadius = 20
    }
    
    private var type: CellType = .right {
        didSet {
            let rightHidden = type == .left ? true : false
            
            rightUserLabel.isHidden = rightHidden;
            rightUserBgView.isHidden = rightHidden;
            rightContentImage.isHidden = rightHidden;
            
            leftUserLabel.isHidden = !rightHidden;
            leftUserBgView.isHidden = !rightHidden;
            leftContentImage.isHidden = !rightHidden;
        }
    }
    
    private var user: String? {
        didSet {
            switch type {
            case .left:  leftUserLabel.text = user
            case .right: rightUserLabel.text = user
            }
        }
    }
    
    func update(type: CellType, message: Message) {
        self.type = type
        self.user = message.userId
        
        
        
        // complete down
        if(self.mediaId != nil && self.mediaId == message.mediaId && self.imageData != nil){
            self.updateImageData(imageData: self.imageData!)
            return
        }
        
        // downing
        if(self.mediaId != nil && self.mediaId == message.mediaId && self.requestId > 0){
            return
        }
        
        self.mediaId = message.mediaId;
        
        // stop last download
        if(self.requestId > 0){
            AgoraRtm.kit?.cancelMediaDownload(self.requestId, completion: nil)
        }
        
        self.requestId = 0
        if let imageData = message.thumbnail {
            self.updateImageData(imageData: imageData)
        }
        
        if let currentMediaId = message.mediaId {
            AgoraRtm.kit?.downloadMedia(toMemory: currentMediaId, withRequest: &self.requestId, completion: {[weak self] (requestId, data, errorCode) in
                
                guard let `self` = self else {
                    return
                }
                self.requestId = 0
                
                guard let `data` = data else {
                    return
                }
                if(errorCode == .ok) {
                    self.updateImageData(imageData: data)
                }
            })
        }
    }
    
    fileprivate func updateImageData(imageData: Data) {
        self.imageData = imageData
        if let image = UIImage(data: imageData) {
            switch (self.type) {
            case .left:
                self.leftContentImage.isHidden = false
                self.leftContentImage.image = image
                break
            case .right:
                self.rightContentImage.isHidden = false
                self.rightContentImage.image = image
                break
            }
        }
    }
}
